src/librustc_resolve/lib.rs

```diff
@@ -2199,7 +2211,30 @@ impl<'a> Resolver<'a> {
                     } else if i == 0 {
                         (format!("use of undeclared type or module `{}`", ident), None)
                     } else {
-                        (format!("could not find `{}` in `{}`", ident, path[i - 1].ident), None)
+                        let mut msg =
+                            format!("could not find `{}` in `{}`", ident, path[i - 1].ident);
+                        if ns == TypeNS {
+                            if let FindBindingResult::Binding(Ok(_)) =
+                                find_binding_in_ns(self, ValueNS)
+                            {
+                                msg = format!(
+                                    "`{}` in `{}` is a concrete value, not a module or Struct you specified",
```

Comment:
It'd be even better if we actually got what the `DefKind` of `ident` is and used it's description here instead of "concrete value".

Reply:
+1, there's a standard wording for this kind of errors - `expected description1, found description2 name`, and it can be reported for both `ns == TypeNS` and `ns == ValueNS` cases:
```
fn main() {}

type A = main; // ERROR expected type, found function `main`
```

Ideally, this case should go through the `smart_resolve_path` infrastructure, but that's a pretty big refactoring.
A small patch solution like this should be acceptable in the meantime even if increases technical debt.